Standardize physics restart behavior

GiudGiud opened the pull request:
This should make @tanoret 's restart be handled entirely by the Physics

refs #28730 

It might be possible to move some code to PhysicsBase.C, as the technique is somewhat the same. but as you can see in WCNSFVTurbulence and WCNSFVScalarPhysics, there are some nuances. So 2/5 cases would likely not use the standard version..

GiudGiud committed:
Add restart capability for turbulence models in physics
- always obey user specification of ICs
- only skip ICs if user specified restart-from-file or not defining variables
- skip default ICs (not user specified) if restarting
refs #28730

GiudGiud committed:
Add comment for scalar variable ICs
refs #28730

GiudGiud committed:
Add advice for handling initial conditions in the Physics documentation
refs #28730

GiudGiud committed:
Skip default ICs and ICs setting in heat conduction physics if user specified to load variables from file
refs #28730

GiudGiud committed:
Add full suite of test for expected behavior of restart + HC physics
refs #28730

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2476207/), step Docs: sync website on cfece5f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28736/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2477046/), step Generate coverage on a8db3f1 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>bfd459</th>
      <th colspan=4>#28736 a8db3f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.01%</td>
      <td>89.02%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4275</td>
      <td>4279</td>
      <td>+4</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>528</td>
      <td>528</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28736/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28736/coverage/heat_transfer)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>bfd459</th>
      <th colspan=4>#28736 a8db3f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.59%</td>
      <td>84.57%</td>
      <td>-0.02%</td>
      <td>82.61%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>16765</td>
      <td>16773</td>
      <td>+8</td>
      <td>19</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3055</td>
      <td>3061</td>
      <td>+6</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28736/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28736/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28736/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 82.61% is less than the suggested 90.0%



This comment will be updated on new commits.

grmnptr reviewed (changes requested):
Very nice! Could you please add a test for the NS module too?

GiudGiud committed:
Add a test for restarting NS physics (all together)
refs #28730

GiudGiud committed:
Default to not linearizing one of the main kernels of K-epsilon when using Newton
refs #9007

GiudGiud committed:
Add turbulent to schmidt number parameter
Avoid overloading console on 0 turbulent viscosity (due to no initialization)
Add support for k-epsilon + scalar advection with a diffusion term

GiudGiud committed:
Add test to ensure desired ICs behavior for physics for:
- flow
- energy
- turbulence
- scalars
refs #28730

GiudGiud committed:
Bound k-epsilon viscosity directly rather than bounding epsilon

GiudGiud committed:
Sparsity pattern adjustments for Newton methods
refs #9007 #23943

GiudGiud commented:
Done. I ended up fixing a few other things that came out. Not initializing the variables is really a terrible idea.

grmnptr commented on modules/navier_stokes/src/functormaterials/INSFVkEpsilonViscosityFunctorMaterial.C:
```diff
@@ -41,7 +41,9 @@ INSFVkEpsilonViscosityFunctorMaterial::INSFVkEpsilonViscosityFunctorMaterial(
                              {
-                               return _C_mu(r, t) * _rho(r, t) * Utility::pow<2>(_k(r, t)) /
-                                      std::max(NS::mu_t_low_limit, _epsilon(r, t));
+                               return std::max(NS::mu_t_low_limit,
```
@tanoret , do you think this change makes sense? To always include a little viscosity?

GiudGiud replied:
I think that was the initial intent here, why else why that variable named like that?

grmnptr commented on modules/navier_stokes/src/fvkernels/INSFVTurbulentDiffusion.C:
```diff
@@ -18,18 +18,22 @@ INSFVTurbulentDiffusion::validParams()
     _scaling_coef(getFunctor<ADReal>("scaling_coef")),
-    _wall_boundary_names(getParam<std::vector<BoundaryName>>("walls"))
+    _wall_boundary_names(getParam<std::vector<BoundaryName>>("walls")),
+    _newton_solve(getParam<bool>("newton_solve"))
```
I think we could decide this automatically by casting the executioner.

GiudGiud replied:
To either steady, transient, multisystem steady etc. Seems easier to just ask the user

grmnptr replied:
No, if it is not inheriting from segregated solver base.

GiudGiud replied:
I created https://github.com/idaholab/moose/issues/28814
I ll fill it up more later.

I dont think a cast is a viable solution, it would not catch other executioners in downstream applications.

grmnptr replied:
I am just saying that right now anything else that is not inheriting from segregated solver base is newton based. Especially for the physics this kernel is supposed to add. If you don't think casting is the right way that is fine with me. Then please add another method to check if the user got this parameter right. I don't want them using the newton solver options with SIMPLE and vice versa.

GiudGiud replied:
Seems like I ll solve #28814 here then

grmnptr commented on modules/navier_stokes/src/fvkernels/INSFVTurbulentDiffusion.C:
```diff
@@ -54,12 +58,21 @@ INSFVTurbulentDiffusion::computeQpResidual()
+    {
+      if (!_newton_solve)
+        return 0;
+      else
+        // we return 0 but preserve the sparsity pattern for Newton's method
+        return 0 * (coeff_elem + coeff_neighbor) *
+               (_scaling_coef(elemArg(), state) + _scaling_coef(neighborArg(), state)) * dudn;
```
This is a little awkward in a user-facing code. Why don't we need this in the regular diffusion kernel? Is this a fix for the restart?

GiudGiud replied:
we actually do from this PR onward, see FVDiffusion

grmnptr commented on modules/navier_stokes/test/tests/finite_volume/physics/restart/2d_channel_scalar_turbulence_init.i:
```diff
@@ -0,0 +1,278 @@
+
+[Problem]
+  solve = false
+  # Currently necessary when using Newton's method, somewhere in the K-epsilon equations we do not preserve
+  # the sparsity pattern when no initial conditions for velocity are used
+  error_on_jacobian_nonzero_reallocation = false
```
You get an error saying that the sparsity pattern is too small? It is a serious bug. That might be the reason why we cannot do split mesh either. We have an option to adjust the pattern, right? Maybe that would be a solution. I know fixing this is not the purpose of this PR, but I wouldn't like this to make it in like this. I can help debug this.

GiudGiud replied:
These sparsity pattern issues got in in the first K-epsilon merge, I am slowly fixing them. I am almost done except for that last case with the 0 velocity initialization, which I don't think it s important to fix since one can just set non-zero velocities to get around.

grmnptr replied:
This is very important to fix. I'll take a look today. I really don't think we use AD information to define the sparsity pattern. Those should be taken care of by the relationship managers. I will have to see those before I believe that by default we allow this behavior.

grmnptr replied:
I'll also tag @lindsayad in case I am too slow finding the answer.

grmnptr replied:
Sorry, I didn't see your comment before. I am certain that the RMs are used for the sparsity pattern. I confirmed it with @roystgnr . I am checking where this can go wrong on the physics side. It is also suspicious that FVDiffusion needs the modifications too. It is the only one that requires `ghost_layers=2+` instead of 1. My suspicion is that something is off with the RM assignment at physics-based restart. I'll come back with actual details.

GiudGiud replied:
I don't think so. First because the timing is similar to what the action was using. 

And then because the ghosting works fine for distributed cases. 

As I mentioned earlier, the new situation is that we end up calling these kernels with a 0 coefficient in the first iteration. Because if you don't initialize the turbulent viscosity that is what you get in your turbulent diffusion terms (momentum and scalar)

GiudGiud replied:
Then in subsequent iterations we stop calling with a 0 coefficient.

grmnptr replied:
Okay we don't use any AD-related information for the sparsity pattern it is defined using the relationship managers alone. Based on this, I think what happens is: we allocate the matrix with a given sparsity pattern, then we don't populate some of the 0 entries and Petsc eliminates the unused entries. This is the default petsc behavior. Will be very hard to work around this besides the changes you added.

GiudGiud replied:
There must be something special using AD though because we are still sending in 0s when we do these little multiplications by 0

grmnptr replied:
No, the zeros are eliminated after assembly:
https://petsc.org/main/manualpages/Mat/MatEliminateZeros/
This is called which gets rid of every offdiagonal zero. 
I am still looking into the details for the precise call stack to make sure we know what is happening.

grmnptr replied:
Never mind, was looking at the wrong call stack. In theory, petsc should shrink the matrix only if we don't put anything in it. That's why putting 0s explicitly works. Will get the call stack soon.

GiudGiud replied:
Once we're done here, let's add some documentation in the AD page and the relationship manager page.

grmnptr replied:
It is actually here:
https://petsc.org/release/manualpages/Mat/MatAssemblyBegin/
And the end routine too.
Actually, Fande added an option to restore the original preallocation before the next assembly.
https://petsc.org/release/manualpages/Mat/MatResetPreallocation/
Maybe we can rely on this when we restart. Let me check if we can use it.

GiudGiud replied:
this code is not related to restarts, but rather to not initializing mu_t

grmnptr replied:
But allowing nonzero jacobian entry allocations on the fly will have a huge burden on computation time and should be avoided if possible. I am checking if it is possible.

GiudGiud replied:
kk well checkout kernels with velocity & k / epsilon.
It's a long task, there are tons of BCs

lindsayad replied:
`MatResetPreallocation` is a great option if the sparsity pattern determined by the relationship managers is always a superset of the true nonzeroes and is not overly dense. This is generally not the case for things like mechanical contact for which the mesh deformation leads to the true sparsity pattern changing over the course of the simulation in which case your initial matrix preallocation would have to be very dense.

lindsayad replied:
When does a Jacobian get assembled when `solve = false`? It would be good to know whether you are getting a new nonzero for the initial Jacobian assembly or a subsequent one. If the first, then there is an issue with the relationship managers. I know that your Navier-Stokes physics adds 3 ghosting layers (corresponding to the `INSFVMixingLengthReynoldsStress` kernel). And I also know that different turbulence inputs require 5-7 ghosting layers perusing the input files.

If you are getting a new nonzero during a Jacobian assembly later than the first, then we either need to reset the preallocation like @grmnptr brought up or we need to ensure we are inserting zeroes during our Jacobian assembly where zeroes may arise in later Jacobian evaluations, e.g. for upwind schemes if the wind changes direction

grmnptr replied:
Okay, how about this:
- I think the ghost layers are probably fine, I think the inputs that require more are the two-phase models.
- Instead of using the early return just use the interpolation below that. I think we ran into trouble because of this early return exception. That will give you the proper derivatives no matter what. You can fall back to average if harmonic is doomed to fail.

If it still errors without `error_on_jacobian_nonzero_reallocation = false` we have a serious problem. 
I hope we won't have to touch the preallocation part. 


Unrelated, but ghosting layer of 7 looks really fishy to me. Feels like an infinite loop and ghosting layer 5-7 was the first that passed because everything was ghosted at that point. Looking at the input we have 4x10 elements in the mesh so it is very likely that we ghost everything. So we might have a big issue for the two-phase models as well.

GiudGiud replied:
> When does a Jacobian get assembled when solve = false?

It does not, I just made sure the input ran if a user wanted to drop the solve = false and run it. 

>  or a subsequent one.

then.

This is simply because of bad zero ICs. I already took care of what mattered here, having a constant sparsity pattern with initialization and with K-Epsilon. Let's remember that the many inputs of ours (iirc msfr) do not run with 0 initial conditions for velocity. So not initializing (to 1e-8 for example) is not viable anyway. 

I know I did not check every single BC in K-epsilon but that's it

>  And I also know that different turbulence inputs require 5-7 ghosting layers perusing the input files.

yeah we should look into it at some point. Is it for turbulence though? I see 3 in the WCNSFV Turbulence physics. I think the 2 phase mixtures are affected. I just checked 2 k-epsilon inputs and they dont have 5-7

> we need to ensure we are inserting zeroes during our Jacobian assembly where zeroes may arise in later Jacobian evaluations

this is what I do here, to go around the user not initializing mu_t. 
If I dont do this, and the user makes an input with kernels, then they will hit the error. With Physics we could enforce that they initialize if we wanted

GiudGiud replied:
> Instead of using the early return just use the interpolation below that. I think we ran into trouble because of this early return exception. That will give you the proper derivatives no matter what. You can fall back to average if harmonic is doomed to fail.

this crashes with harmonic.
I don't see how
```
if (coeff. == 0)
  do_average_interpolation
else
  do_interpolation_user_requested
```
is any better than what I have now. 

> If it still errors without error_on_jacobian_nonzero_reallocation = false we have a serious problem.

requiring users to set their initial velocities to 1e-8 is not a serious problem. 
If you have an example where things are initialized properly and we get errors, then I agree

grmnptr replied:
The issue is that we have these min max limiters here and there (including the early return). And those don't carry the derivatives. They are just numbers. So we have big differences between existing derivatives depending on the raw value. And yes, there is a big difference in terms of logic. That way this sparsity pattern preservation issue does not get exposed in user-facing functions. I am really second guessing that we should allow k-epsilon with Newton.

grmnptr replied:
In my opinion, users should not be aware of the sparsity pattern and should not learn about it either. I think it would be bad design if they had to worry about it in an fvdiffusion just because of the limiters in k epsilon discarding derivatives left and right.

grmnptr replied:
Also the fvdiffusion qp function is in the moose worksop slide and I think showcasing this with this sparsity pattern hackery would be embarassing compared to the fekernel for example.

GiudGiud replied:
But it is a user facing problem. If you are looking at this code, it's because you are most likely coding a diffusion kernel. And if you are doing so and are not mindful of the sparsity pattern, then you ll get an error. 

if we go hide the multiplication by 0 or a sneaky switch to average inside the MathFVUtils.h we leave the user open to those issues. 

>  I am really second guessing that we should allow k-epsilon with Newton.

It has been allowed since it was merged. Its sparsity pattern issues are essentially fixed and there's even an action for it now. Seems like not the time to rip it out.

GiudGiud replied:
I'm totally ok with a disclaimer that we dont recommend using it and that we are working on something better.

GiudGiud replied:
> In my opinion, users should not be aware of the sparsity pattern and should not learn about it either.

at some point they do though. Coding in FV they will inevitably hit the issue.

>  I think it would be bad design if they had to worry about it in an fvdiffusion just because of the limiters in k epsilon discarding derivatives left and right.

again not that. mu_t = 0 is the source here.

grmnptr replied:
Okay how about:
- we remove this from fvdiffusion so that at least that is presentable and does not do it for this kepsilon solution alone
- make sure that we can run without the `allocation=false` option. 
I have a few ideas, will try them tomorrow. Will push a commit or 2.

GiudGiud replied:
but you're right that there is some limiting in the functor material route. So maybe that is affected too. 
but again, mu_t = 0 is the route problem, which comes from users not initializing tke

> Also the fvdiffusion qp function is in the moose worksop slide and I think showcasing this with this sparsity pattern hackery would be embarassing compared to the fekernel for example.

I guess that's a good reason. It s too early for the simple FV tutorial. 
I ll think of something.

grmnptr replied:
I think adding the interpolation with an average override if harmonic is in trouble will solve the fvdiffusion side. I can hunt down some derivatives tomorrow.

GiudGiud replied:
you sure your time would not be better spent writing some linear kernels for this :D ?

GiudGiud replied:
hunting down derivatives for the theoretical case of users yolo-ing the initial conditions would not be my go to

grmnptr replied:
I think the fix might be actially not that hard. We should probably just add some derivatives manually when we limit. But yeah is that a nicer solution than this, most likely not. Ah man.

GiudGiud replied:
to solve mu_t = 0 (true 0), let s just run the auxkernel on initial and add a limiter to it
then we can take off this logic

grmnptr replied:
Oh wait, I thought the one with newton's method used a functormaterial for mut. I'll check the implementation, this whole thread could have been a misunderstanding.

GiudGiud replied:
ah it is
```
        mu_t_as_aux_variable = false
```

GiudGiud replied:
interesting. 
Maybe it was hitting a zero mu_t (instead of a 1e-8) before I moved the limiter from the denominator to the numerator

GiudGiud replied:
Or maybe it is the limiter then. 
So off of the lack of initialization of the tke.

GiudGiud replied:
But regular harmonic interpolation would have trouble with it too. The early return isn't triggered when the viscosity is limited

grmnptr replied:
I think in that case the issue is that we won't have the derivatives of k and epsilon within mu_t in the first assembly. 
I'll take a look. Was it a hard requirement to not initialize tke?

lindsayad replied:
In an ideal world users should not have to think about sparsity pattern. They have not had to do so for the first 16 years of MOOSE.

I'm not saying that it's not good for them to think about it. I'm just saying ideally they shouldn't *have* to

lindsayad replied:
Like there was a time users had to set non-zero velocity for INSAD (and maybe for FV) to avoid undefined derivatives due to taking the norm of the velocity. I thought that didn't look very nice in the input so I slowly fixed that

GiudGiud replied:
the lid-driven energy case also works now with no initial conditions

grmnptr reviewed (changes requested):
The issue with the sparsity pattern is very concerning.

GiudGiud commented on framework/src/fvkernels/FVDiffusion.C:
```diff
@@ -54,8 +54,9 @@ FVDiffusion::computeQpResidual()
     // have a harmonic interpolation)
+    // We multiply by 0 to keep the sparsity pattern
     if (!coeff_elem.value() && !coeff_neighbor.value())
-      return 0;
```
I think on the first calculation we examine the sparsity pattern then try to keep the same. 
If the first calculation has 0 viscosity (like k-epsilon with no initialization of mu_t), then you get the wrong sparsity pattern

grmnptr replied:
I think I'll have to look at the ghosting functors for this.

grmnptr replied:
I would like to believe that we know the sparsity pattern before the matrices are allocated for the very first time. But I'll check.

GiudGiud replied:
how would we though? like how could it know that D depends on temperature?

GiudGiud committed:
Add option to preserve sparsity pattern at the problem level
Document all 3 options for fine tuning this behavior
Set segregated solvers not to preserve it
refs #28814

GiudGiud committed:
Use new sparsity pattern API in Navier Stokes when the only arrangements for
the turbulence objects were made for sparsity pattern preservation
refs #28814

GiudGiud committed:
Add test to make sure we no longer need to error on jacobian re-allocation for k-epsilon with 0 initial velocity
refs #28814

GiudGiud commented on modules/navier_stokes/src/fvkernels/INSFVTurbulentAdvection.C:
```diff
@@ -53,7 +57,10 @@ INSFVTurbulentAdvection::computeQpResidual()
                              determineState());
-  return _normal * MetaPhysicL::raw_value(v) * rho_face * var_face;
+  if (_newton_solve)
+    return _normal * v * rho_face * var_face;
```
I think here it s more about keeping the derivatives (which possibly improved out Newton convergence, but I did not really see it get better) than preserving the sparsity pattern

GiudGiud replied:
>  What would your opinion be on separating kernels and materials for segregated solves and Newton solves?

This is where we are going in a way already. But we need to contribute the linearFV version of these kernels.
Otherwise wrt the nonlinear kernel, no I dont want to duplicate roughly the same thing. Especially since I am trying to not have fvkernels be user facing anymore with these physics

GiudGiud replied:
How would having duplicated kernels for nonlinear FV Newton and SIMPLE be easier to maintain?

grmnptr replied:
Yeah I think when I wrote this down I didn't consider that hopefully the segregated functionalities will migrate to linearFV and then we'll have no trouble maintaining shared kernels between two very different solution algorithms.

GiudGiud committed:
Add option to neglect derivatives on the TKE/TKED

GiudGiud committed:
Add Jacobian testing for K-Epsilon + energy

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2494619/), step Conda libmesh build config check on cbb4354 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

grmnptr commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -464,6 +464,7 @@ FEProblemBase::FEProblemBase(const InputParameters & parameters)
             ? getParam<bool>("error_on_jacobian_nonzero_reallocation")
             : _app.errorOnJacobianNonzeroReallocation()),
     _ignore_zeros_in_jacobian(getParam<bool>("ignore_zeros_in_jacobian")),
+    _preserve_matrix_sparsity_pattern(true),
```
If ignore-zeros is a user-facing option, I would like this to be user-facing too. Also, I think for FV, depending on the coupling, many entries get condensed out. For example, if the material properties depend on cell temperature alone then all the coupling with the neighbor temperatures will be condensed out. Also, this could be really confusing, if someone notices that this is true and petsc still condenses the entries out. Plus, this is FV specific at the moment. I think maybe adding some documentation to this to a design page would make these clear.

GiudGiud replied:
why? do you have a use for it? Having parameters we dont use does not help anyone.

GiudGiud replied:
Especially since other kernels might not be implementing this correctly at this point. It's not enforced, we can have this set to true and kernels from 3 years ago don't preserve sparsity pattern

grmnptr replied:
Okay, I suppose this is temporary in a sense that when we migrate the turbulence and other routines to the linear solver, we can remove this duality in the nonlinear FV kernels. And then we might be able to remove this too.

GiudGiud replied:
I'd be fine with removing it if the need goes away.

grmnptr reviewed (changes requested):
Looking nice, just a few more questions/comments.

GiudGiud commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1881,6 +1881,22 @@ class FEProblemBase : public SubProblem, public Restartable
+   */
+  bool preserveMatrixSparsityPattern() const { return _preserve_matrix_sparsity_pattern; };
+
+  void setPreserveMatrixSparsityPattern(bool preserve);
```
```suggestion

  /// Set whether the sparsity pattern of the matrices being formed during the solve (usually the Jacobian)
  /// should be preserved.
  void setPreserveMatrixSparsityPattern(bool preserve);
```

GiudGiud commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1881,6 +1881,22 @@ class FEProblemBase : public SubProblem, public Restartable
+   */
+  bool ignoreZerosInJacobian() const { return _ignore_zeros_in_jacobian; }
+
+  void setIgnoreZerosInJacobian(bool state) { _ignore_zeros_in_jacobian = state; }
```
```suggestion
  /// Set whether the zeros in the Jacobian should be dropped from the sparsity pattern
  void setIgnoreZerosInJacobian(bool state) { _ignore_zeros_in_jacobian = state; }
```

GiudGiud committed:
Add docstrings for two problem attribute setters

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2505216/), step Clang format on 45cb94c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28736/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28736/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b3924ece4ed556a7ca9811a6e046eb5c4503df72`

grmnptr reviewed (approved):
This is very nice functionality! Failures look unrelated to me.

GiudGiud commented:
Thanks for the review!
@tanoret fix is in for restarts if you want to update your inputs
